File: py/D8.py
from dataclasses import dataclass
import copy

class op:
    JMP = "jmp"
    NOP = "nop"
    ACC = "acc"

@dataclass
class instruction:
    op: op
    arg: int

def parseInput():
    file = open("../inputs/D8.input", "r")
    program = []
    contents = file.read()
    for line in contents.splitlines():
        split = line.split()
        program.append(instruction(split[0], int(split[1])))
    return program

def run(program):
    seen = set()
    pptr = 0
    acc = 0
    while True:
        if pptr in seen: return acc, False
        if pptr >= len(program): return acc, True
        seen.add(pptr)
        
        instruction = program[pptr]

        if instruction.op == op.JMP:
            pptr += instruction.arg
        elif instruction.op == op.ACC:
            acc += instruction.arg
            pptr += 1
        elif instruction.op == op.NOP:
            pptr += 1

def p1():
    acc, halts = run(parseInput())
    return acc

def p2():
    program = parseInput()
    for i in range(len(program)):
        modified = parseInput()
        instruction = modified[i]
        
        if instruction.op == op.JMP:
            instruction.op = op.NOP
        elif instruction.op == op.ACC:
            continue
        elif instruction.op == op.NOP:
            instruction.op = op.JMP

        acc, halts = run(modified)

        if halts: return acc

print(p1())
print(p2())

File: rb/D8.rb
require "set"

$JMP = "jmp"
$NOP = "nop"
$ACC = "acc"

Instruction = Struct.new(:op, :arg)

def parseInput()
  program = []
  File.open("../inputs/D8.input", "r") do |f|
    for line in f.each_line
      line.chomp =~ /(.{3}) (.+)/
      program.append(Instruction.new($1, Integer($2)))
    end
  end
  return program
end

def run(program)
  pptr = 0
  acc = 0
  seen = Set[]
  while true
    return acc, false if seen.include? pptr
    return acc, true if pptr >= program.length
    seen.add pptr
    case program[pptr].op
    when $NOP
      pptr += 1
    when $JMP
      pptr += program[pptr].arg
      next
    when $ACC
      acc += program[pptr].arg
      pptr += 1
    end
  end
end

def p1()
  return (run parseInput)[0]
end

def p2()
  program = parseInput()
  for i in 0...program.length
    modified = parseInput()
    case modified[i].op
    when $ACC 
      next
    when $JMP 
      modified[i].op = $NOP
    when $NOP 
      modified[i].op = $JMP
    end
    acc, halts = run (modified)
    return acc if halts
  end
end

puts p1()
puts p2()